### DISCUSSION TITLE
i am getting state mismatch in chrome only although the code works fine in firefox

### DISCUSSION THREAD

**Discussion by mahoudahmed** [Azure-Samples/ms-identity-python-webapp#151] [Q&A] [answered]
i am using azure to login to my flask app 

this is my login 
```
@app.route("/login")
def login():
    login_params = auth.log_in(
        scopes=app_config.SCOPE,
        redirect_uri=url_for("auth_response", _external=True)
    )
    return render_template("login.html", **login_params)
```

and this is auth_response to be executed after login 

```
@app.route(app_config.REDIRECT_PATH)
def auth_response():
    result = auth.complete_log_in(request.args)
    if "error" in result:
        return render_template("auth_error.html", result=result)
    return redirect(url_for("index"))
```

in auth_response i am getting state mismatch in chrome only but works fine in firefox

**Comment by rayluo** [comment, marked as answer]
Your code snippet was derived from a very old version of our sample. Please take a look at [the latest implementation](https://github.com/Azure-Samples/ms-identity-python-webapp); your first reaction now may be reluctant to start over, but that is likely easier than debugging your old app, because the new usage pattern is vastly simplified. Please reply back if you still run into any issues.

**Reply by mahoudahmed** [reply]
thanks for your help , i updated the library to the latest and i followed the steps 

i used redis to store the session 
```
#app.py
app.wsgi_app = ProxyFix(app.wsgi_app, x_proto=1, x_host=1)
app.jinja_env.globals.update(Auth=identity.web.Auth)
app.config['SESSION_PERMANENT'] = False
app.config['SESSION_COOKIE_SECURE'] = True
redis_url = app.config['CELERY_RESULT_BACKEND']
parsed_url = urlparse(redis_url)
path_str = parsed_url.path.decode('utf-8') if isinstance(parsed_url.path, bytes) else parsed_url.path
db_index = int(path_str.lstrip('/'))

app.config['SESSION_TYPE'] = 'redis'
app.config['SESSION_REDIS'] = redis.StrictRedis(
    host=parsed_url.hostname,
    port=parsed_url.port,
    db=db_index
)
Session(app)
```

in my login.py 

```
@app.route("/login")
def login():
    return render_template("login.html", **auth.log_in(
        scopes=app_config.SCOPE, 
        redirect_uri=url_for("auth_response", _external=True), 
        ))

@app.route(app_config.REDIRECT_PATH)
def auth_response():
    result = auth.complete_log_in(request.args)
    if "error" in result:
        return render_template("auth_error.html", result=result)
    return redirect(url_for("index"))
```

i am using identity>=0.5.1,<0.6 
now the code still not working 'ValueError: state mismatch: khtFaKRjDcivUwIZ vs TxSNFgwDjHvBuhXJ' in chrome and edge , but works fine in firefox 

also i want to mention that i am using flask with docker and gunicorn with GeventWebSocketWorker for websockets

**Reply by rayluo** [reply]
> > Your code snippet was derived from a very old version of our sample. Please take a look at [the latest implementation](https://github.com/Azure-Samples/ms-identity-python-webapp);
>
> thanks for your help, i updated the library to the latest and i followed the steps
> ...
> i am using identity>=0.5.1,<0.6

You are still using old code snippet dated back to identity 0.5.x era. Have you tried [starting from our sample](https://github.com/Azure-Samples/ms-identity-python-webapp?tab=readme-ov-file#getting-started) with zero modification? Please help me to help you, by following the recommended steps.

**Comment by mahoudahmed** [comment]
thanks

**Comment by mahoudahmed** [comment]
@rayluo  hi , first thanks everything is working fine now , i just have one simple question , now the auth_response function is not being called after login like before , although i put it in redirect_uri
